db/schema_tables.cc

```diff
@@ -354,6 +355,10 @@ schema_ptr scylla_tables(schema_features features) {
             sb.with_column("partitioner", utf8_type);
             offset |= 0b10;
         }
+        if (has_group0_schema_versioning) {
+            sb.with_column("originated_from_group0", boolean_type);
```

Comment:
We should avoid extending the schema for such corner cases. Is it not enough for recovery to create a new schema version once it chooses a leader? I'm not familiar with recovery.

Reply:
There is no leader in recovery. Recovery means we're operating without group 0.

But recovery is not the only reason why we need this. Another is that we might have upgraded a cluster where some tables were created using the old way. We need to distinguish between old-way tables (for which we calculate digests) and tables created through group 0 (where version is pre-determined).

Reply:
We don't elect a leader inside recovery mode because there is no group 0, no Raft. Recovery is pretty much the state before upgrade.

Reply:
But why do we need to calculate digests? We can just pause all of that until we re-form group 0.

Imagine if we started the project from scratch, we wouldn't have added schema digests just for that.

Reply:
You mean we should disable schema changes completely in recovery mode?

Reply:
Yes, focus on recovery not on adding that column you always wanted.

Reply:
I imagine that doing as schema change might be necessary part of recovery, because we're in some broken state and triggering a schema change is a way out.

On the other hand, these situations should be rare, and we should be able to still solve them by directly modifying system tables if necessary. So maybe we can disable schema changes in recovery.

Still, for now I don't see how to deal with upgrades without this column or something similar. Maybe in the future upgrades from pre-group0 clusters will disappear, but today, they exist, which means we must assume that some tables have been created in the old mode. Even if I added some kind of schema rewrite procedure (which would complicate the whole solution), we need to deal with these schemas before the rewrite happens, during upgrade and so on.

Reply:
Ok, but let's call it something else. originated_from_group0 doesn't mean anything to people not in the know.

Reply:
This name describes exactly the value stored by this column, nothing less, nothing more. You won't understand what this column does if you don't know what group 0 is, no matter how you name it, and I believe that any name which doesn't mention "group 0" would be strictly less descriptive. How would you name it?

Reply:
Maybe "committed_by_group0".

Reply:
Or "committed_by_raft" / "committed_through_raft" if we're really scared of mentioning "group 0". (Though I think that "group 0" is more accurace than using "raft".)

Reply:
group0 is fine. but what does "originated_from_group0" mean?

Reply:
That group 0 is the origin (source) of this particular schema definition, i.e. it was committed through group 0, as opposed to the old way of performing schema changes, where we save the definition in a local table and then send it directly to other nodes (and if that fails, it is eventually disseminated through gossiping.)

Reply:
FWIW I agree with @kbr-scylla that not supporting schema changes in recovery mode will greatly reduce its usefulness. There is even an argument in "maintenance mode" PR that the mode should allow schema changes and at this point a cluster (or rather its single node) is in much more degraded state. May be if we would have written everything from scratch we would not write this digest code , but it is there, so why not use it. We still can deprecate and remove it in two years if we will see that having DDL in recovery mode is not useful. We will have an experience by then.

Reply:
RECOVERY mode would also look completely differently if we started from scratch.

Right now, RECOVERY mode works by reverting to the gossiper-based way of things. There we can do cluster operations in the old ways, including schema changes. Going back to group 0 is just the regular upgrade-to-raft procedure.

If we did things from scratch, we would basically need an external orchestration tool to perform recovery, remove any dead nodes and re-establish group 0.

Reply:
I opened a new issue so we don't forget about this.
https://github.com/scylladb/scylladb/issues/15422